### PR TITLE
Perform dereference pushdown after table scan redirection

### PR DESCRIPTION
**Pull request by raunaqmorarka**

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/sql/planner/PlanOptimizers.java [hunk 1]
```diff
@@ -358,7 +358,6 @@ public PlanOptimizers(
                 new PruneWindowColumns());
 
         Set<Rule<?>> projectionPushdownRules = ImmutableSet.of(
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/sql/planner/PlanOptimizers.java:360]: what if we keep the code here unchanged and just add rules
```
                new PushProjectionThroughUnion(),
                new PushProjectionThroughExchange(),
```
explicitly at
`io/trino/sql/planner/PlanOptimizers.java:438`
Then I think we don't have to change anything else in `PlanOptimizers`

File: core/trino-main/src/main/java/io/trino/sql/planner/PlanOptimizers.java [hunk 2]
```diff
@@ -384,12 +383,6 @@ public PlanOptimizers(
                         new InlineProjections(),
                         new RemoveRedundantIdentityProjections()));
 
-        IterativeOptimizer projectionPushDown = new IterativeOptimizer(
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/sql/planner/PlanOptimizers.java:387]: undo. Let's keep it where it is for consistency for now

File: testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java [hunk 3]
```diff
@@ -89,10 +91,11 @@ protected QueryRunner createQueryRunner()
             public Iterable<ConnectorFactory> getConnectorFactories()
             {
                 MockConnectorFactory connectorFactory = MockConnectorFactory.builder()
-                        .withListTables((session, s) -> ImmutableList.of(new SchemaTableName("default", "test_table")))
+                        .withListTables((session, s) -> ImmutableList.of(new SchemaTableName("default", "tests_table")))
```
Comment (sopel39) [inline testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java:94]: undo? why these changes? seem unrelated

**Review by sopel39 — COMMENTED** [review]
small comments

File: testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java (hunk 3 above)
Comment (raunaqmorarka) [reply testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java:94]: They are not unrelated, `TestEventListenerBasic#testPlanningFailure` starts failing after the first PushProjectionIntoTableScan is moved.

File: core/trino-main/src/main/java/io/trino/sql/planner/PlanOptimizers.java (hunk 1 above)
Comment (raunaqmorarka) [reply core/trino-main/src/main/java/io/trino/sql/planner/PlanOptimizers.java]: It solves the problem for table scan redirection but I'm not sure whether it breaks anything for dereference pushdown

File: testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java (hunk 3 above)
Comment (sopel39) [reply testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java:94]: We do we need to change `test_table` to `tests_table`?
Why do we need `getColumns`?

File: testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java [hunk 4]
```diff
@@ -144,7 +147,7 @@ public void testParseError()
     public void testPlanningFailure()
             throws Exception
     {
-        assertFailedQuery("SELECT * FROM mock.default.tests_table", "Throw from apply projection");
+        assertFailedQuery("SELECT test + 1 FROM mock.default.tests_table", "Throw from apply projection");
```
Comment (sopel39) [inline testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java:150]: Why it does not fail with `*`?

File: core/trino-main/src/main/java/io/trino/sql/planner/PlanOptimizers.java [hunk 5]
```diff
@@ -435,8 +435,9 @@ public PlanOptimizers(
                         estimatedExchangesCostCalculator,
                         ImmutableSet.<Rule<?>>builder()
                                 .addAll(columnPruningRules)
-                                .addAll(projectionPushdownRules)
                                 .addAll(ImmutableSet.of(
+                                        new PushProjectionThroughUnion(),
```
Comment (phd3) [inline core/trino-main/src/main/java/io/trino/sql/planner/PlanOptimizers.java:439]: my understanding is that we want to converge towards being able to put all rules into this particular iterative optimizer. so I'm not sure if removing stuff from here is the right direction. 

The core issue the PR is trying to solve seems to be that connector pushdown support may be different in source and destination, so may be it's better not to change anything other than `PushProjectionIntoTableScan` invocation. 

@martint wdyt?

File: core/trino-main/src/main/java/io/trino/sql/planner/PlanOptimizers.java [hunk 6]
```diff
@@ -568,6 +569,7 @@ public PlanOptimizers(
 
         // Perform redirection before CBO rules to ensure stats from destination connector are used
         // Perform redirection before agg, topN, limit, sample etc. push down into table scan as the destination connector may support a different set of push downs
+        // Perform redirection before push down of dereferences into table scan via PushProjectionIntoTableScan
         // Perform redirection after at least one PredicatePushDown and PushPredicateIntoTableScan to allow connector to use pushed down predicates in redirection decision
```
Comment (phd3) [inline core/trino-main/src/main/java/io/trino/sql/planner/PlanOptimizers.java:573]: (not in scope of this PR) it seems a bit arbitrary to me to let the destination table to be aware of predicate pushdown, but not other pushdowns. in long term, would it be better to let the `applyTableScanRedirect` API return equivalents of "residues" for all the pushdowns that it doesn't understand - and recreate those nodes in the engine?

Comment (raunaqmorarka) [reply core/trino-main/src/main/java/io/trino/sql/planner/PlanOptimizers.java:573]: You are right that we could potentially extend applyTableScanRedirect to "reverse" every type of pushdown. But the implementation becomes more complex with every type of pushdown we try to deal with.
The awareness of pushed down predicate was added as it could potentially be a factor in the redirection decision (there may be multiple candidate destinations and we might choose between them based on the predicate).

File: core/trino-main/src/test/java/io/trino/sql/planner/TestTableScanRedirectionWithPushdown.java [hunk 7]
```diff
@@ -268,26 +308,45 @@ else if (name.equals(DESTINATION_TABLE)) {
...
+                newVariableName = ((Variable) projection).getName();
+                newColumnHandle = assignments.get(newVariableName);
+            }
+            else if (projection instanceof FieldDereference) {
+                FieldDereference dereference = (FieldDereference) projection;
+                if (!(dereference.getTarget() instanceof Variable)) {
+                    throw new UnsupportedOperationException();
+                }
+                newVariableName = ((Variable) dereference.getTarget()).getName() + "#" + dereference.getField();
+                newColumnHandle = new MockConnectorColumnHandle(newVariableName, projection.getType());
```
[the remark this answers is not in the document]
Comment (raunaqmorarka) [reply core/trino-main/src/test/java/io/trino/sql/planner/TestTableScanRedirectionWithPushdown.java:329]: changed now

File: testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java (hunk 3 above)
Comment (raunaqmorarka) [reply testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java:94]: `test_table` is wrong as the query in test uses `tests_table `
We need some non-trivial projection on an actual column now to avoid the ProjectNode from getting pruned before PushProjectionIntoTableScan has a chance to execute.

File: testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java (hunk 4 above)
Comment (raunaqmorarka) [reply testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java:150]: With `*` PushProjectionIntoTableScan doesn't run, it seems the projection gets pruned before that rule gets a chance to run.

File: core/trino-main/src/main/java/io/trino/sql/planner/PlanOptimizers.java (hunk 5 above)
Comment (sopel39) [reply core/trino-main/src/main/java/io/trino/sql/planner/PlanOptimizers.java:439]: The problem is that our optimizer is greedy. So depending on rule execution (e.g pushdown into connector rules), we might not have a chance to apply redirection rule. Therefore, it would be best if we could "try" redirection rule before any pushdowns.

Generally, it's odd that we do connector pushdown that early while any other connector pushdowns (predicate, aggregations, joins) is performed later.

File: core/trino-main/src/main/java/io/trino/sql/planner/PlanOptimizers.java (hunk 6 above)
Comment (sopel39) [reply core/trino-main/src/main/java/io/trino/sql/planner/PlanOptimizers.java:573]: In the future, we should rely on materialized views substituion. Optimizer will be able to match MV vs query subplan and substitute subplan with MV scan. With that solution, there is no problem of "reversing" pushdown.

File: testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java (hunk 4 above)
Comment (sopel39) [reply testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java:150]: would `test` itself be sufficient? Why we need `+1`?

File: core/trino-main/src/main/java/io/trino/sql/planner/PlanOptimizers.java (hunk 5 above)
Comment (raunaqmorarka) [reply core/trino-main/src/main/java/io/trino/sql/planner/PlanOptimizers.java]: Made changes now to just move PushProjectionIntoTableScan

File: testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java (hunk 4 above)
Comment (raunaqmorarka) [reply testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java:150]: even with just `test`, the PushProjectionIntoTableScan doesn't trigger, putting a non-trivial expression in the projection was the only way I was able to get the ProjectNode to stay until PushProjectionIntoTableScan rule is triggered.

**Review by sopel39 — APPROVED** [review]

**Comment by sopel39** [comment]
@martint @phd3 PTAL

**Review by phd3 — APPROVED** [review]
Looks good to me except for one comment.

File: core/trino-main/src/main/java/io/trino/sql/planner/PlanOptimizers.java [hunk 8]
```diff
@@ -579,7 +573,7 @@ public PlanOptimizers(
                         estimatedExchangesCostCalculator,
                         ImmutableSet.of(
                                 new ApplyTableScanRedirection(metadata),
-                                new PushProjectionIntoTableScan(metadata, typeAnalyzer),
+                                new PruneTableScanColumns(metadata),
```
[the remark this answers is not in the document]
Comment (raunaqmorarka) [reply core/trino-main/src/main/java/io/trino/sql/planner/PlanOptimizers.java:576]: Yes, the target table may contain only a subset of the source table's columns.
We should be able to perform redirection when the columns accessed by the table scan are present in the target table.
I've now updated the comment in PruneTableScanColumns

File: core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PruneTableScanColumns.java [hunk 9]
```diff
@@ -42,7 +42,8 @@
 import static java.util.Objects.requireNonNull;
 
 /**
- * TODO: this is a special case of PushProjectionIntoTableScan and should be merged with that rule.
+ * This is a special case of PushProjectionIntoTableScan that performs only column pruning.
+ * ApplyTableScanRedirection needs to run after column pruning but before dereference pushdown.
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PruneTableScanColumns.java:46]: `but before dereference pushdown.` -> `but before more complex projection pushdown`

File: core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PruneTableScanColumns.java [hunk 10]
```diff
@@ -42,7 +42,8 @@
 import static java.util.Objects.requireNonNull;
 
 /**
- * TODO: this is a special case of PushProjectionIntoTableScan and should be merged with that rule.
+ * This is a special case of PushProjectionIntoTableScan that performs only column pruning.
+ * ApplyTableScanRedirection needs to run after column pruning but before more complex projection pushdown.
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PruneTableScanColumns.java:46]: this is out of context. Either improve message or remove this sentence
